### ISSUE TITLE
ud_grade modifies input array

### ISSUE THREAD

**Issue by lpsinger** [healpy/healpy#262]
The `ud_grade` function modifies the input array in some circumstances. This is probably an undesired side effect. The issue occurs if there are any invalid values (`hp.UNSEEN`) in the input array. In this case, the corresponding entries are overwritten with zero in `_ud_grade_core`:

        m[bads] = 0

**Comment by zonca** [comment]
can you provide an example that fails?

`UNSEEN` should be restored back at https://github.com/healpy/healpy/blob/master/healpy/pixelfunc.py#L1664

**Comment by lpsinger** [comment]
Oh, I didn't notice that. Still, it's generally bad practice to modify the input. This will not work for the most general Numpy array. See, for instance, the following example:

```python
>>> import numpy as np
>>> import healpy as hp
>>> m = np.repeat(hp.UNSEEN, 48)
>>> m.flags.writeable = False
>>> hp.ud_grade(m, 1, order_in='NEST', order_out='NEST')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/healpy/pixelfunc.py", line 226, in wrapper
    out = f(m, *args, **kwds)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/healpy/pixelfunc.py", line 1613, in ud_grade
    mout = _ud_grade_core(m,nside_out,pess=pess, power=power, dtype=dtype)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/healpy/pixelfunc.py", line 1650, in _ud_grade_core
    m[bads] = 0
ValueError: assignment destination is read-only
```

**Comment by zonca** [comment]
Any other option without making a copy of the map?

**Comment by lpsinger** [comment]
Not that I can think of.

**Comment by cjcopi** [comment]
Replacing https://github.com/healpy/healpy/blob/master/healpy/pixelfunc.py#L1653 with
```
map_out = np.sum(mr*hit, axis=1).astype(type_out)
```
and removing the writes to `m` would make the function work for read only maps.

**Comment by zonca** [comment]
Great! Could you create a pull request for that?

**Comment by zonca** [comment]
fixed in #264
